src/application/predictor.hpp

```diff
@@ -133,37 +153,64 @@ class Predictor {
     }
-    std::function<double(const std::vector<std::pair<int, double>>&)> predict_fun;
-    if (is_simgoid_) {
-      predict_fun = [this](const std::vector<std::pair<int, double>>& features) {
-        return PredictOneLine(features);
-      };
-    } else {
-      predict_fun = [this](const std::vector<std::pair<int, double>>& features) {
-        return PredictRawOneLine(features);
-      };
+    if (predict_leaf_index) {
+        std::function<std::string(const std::vector<std::pair<int, double>>&)> predict_fun;
+        predict_fun = [this](const std::vector<std::pair<int, double>>& features) {
+            return PredictLeafIndexOneLine(features);
+        };
+        std::function<void(data_size_t, const std::vector<std::string>&)> process_fun =
+          [this, &parser_fun, &predict_fun, &result_file]
+        (data_size_t, const std::vector<std::string>& lines) {
+          std::vector<std::pair<int, double>> oneline_features;
+          std::vector<std::string> pred_result(lines.size(), "");
+    #pragma omp parallel for schedule(static) private(oneline_features)
+          for (data_size_t i = 0; i < static_cast<data_size_t>(lines.size()); i++) {
+            oneline_features.clear();
+            // parser
+            parser_fun(lines[i].c_str(), &oneline_features);
+            // predict
+            pred_result[i] = predict_fun(oneline_features);
+          }
+
+          for (size_t i = 0; i < pred_result.size(); ++i) {
+            fprintf(result_file, "%s\n", pred_result[i].c_str());
+          }
+        };
+        TextReader<data_size_t> predict_data_reader(data_filename);
+        predict_data_reader.ReadAllAndProcessParallel(process_fun);
+    }
+    else {
+        std::function<double(const std::vector<std::pair<int, double>>&)> predict_fun;
+        if (is_simgoid_) {
+          predict_fun = [this](const std::vector<std::pair<int, double>>& features) {
+            return PredictOneLine(features);
+          };
+        } else {
+          predict_fun = [this](const std::vector<std::pair<int, double>>& features) {
+            return PredictRawOneLine(features);
+          };
+        }
+        std::function<void(data_size_t, const std::vector<std::string>&)> process_fun =
+          [this, &parser_fun, &predict_fun, &result_file]
+        (data_size_t, const std::vector<std::string>& lines) {
+          std::vector<std::pair<int, double>> oneline_features;
+          std::vector<double> pred_result(lines.size(), 0.0f);
+    #pragma omp parallel for schedule(static) private(oneline_features)
+          for (data_size_t i = 0; i < static_cast<data_size_t>(lines.size()); i++) {
+            oneline_features.clear();
+            // parser
+            parser_fun(lines[i].c_str(), &oneline_features);
+            // predict
+            pred_result[i] = predict_fun(oneline_features);
+          }
+
+          for (size_t i = 0; i < pred_result.size(); ++i) {
+            fprintf(result_file, "%f\n", pred_result[i]);
```

Comment:
I am actually planning to change return type of PredictLeafIndex into int*. It might be used for python/R wrapper to load results into dataframe more easily.